### PR TITLE
Austin/histogram

### PR DESCRIPTION
**Pull request by AustinBGibbons**
For #3, committing the first step in building a histogram. Takes an Iterable and number of bins (defaulted to 10). For example:

```histogram(Seq(1, 2, 2, 3, 3, 4, 4, 4, 4, 4, 5, 5, 5, 6, 7), 7)```

chart 1
chart 5

### CODE REVIEW

**Comment by squito** [comment]
I think column chart is the right choice.  But I think you should also have a way to plot already binned data.  Eg., say you are computing the histogram in a distributed fashion from GBs of data.  You should also have a:

`def histogram(binnedData:Iterable[(Double,Double,Double)])`

with `(min,max,count)` for each triple, or something like that

**Comment by AustinBGibbons** [comment]
Wise, I think I will further add grouped data `Iterable[Iterable[Numeric]]` 
as well as bucketKey->bucketCount pairs: `Iterable[(Numeric, Numeric)]`

**Comment by squito** [comment]
whats the point of grouped data?  the user could just do a `.flatten`, doesn't save space or time.  Doesn't hurt to add it, though.

I guess in most cases, just a `(key,count)` is good enough, but I often want to control the bucket sizes explicitly.  Often with power law data I make the buckets get bigger as the values get bigger. (eg., 1 to 10, 10 to 100, 100 to 1000).  Kind of a special case, but pretty common in most big data stuff.

**Comment by AustinBGibbons** [comment]
I'm imagining grouped data as a precursor to a key(s): count - ie, you
group data yourself by an unorthodox function (say, such as grouping by
log) and not having to built the countmap.

On Fri, Jan 23, 2015 at 7:47 AM, Imran Rashid <notifications@github.com>
wrote:

> whats the point of grouped data? the user could just do a .flatten,
> doesn't save space or time. Doesn't hurt to add it, though.
>
> I guess in most cases, just a (key,count) is good enough, but I often
> want to control the bucket sizes explicitly. Often with power law data I
> make the buckets get bigger as the values get bigger. (eg., 1 to 10, 10 to
> 100, 100 to 1000). Kind of a special case, but pretty common in most big
> data stuff.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/quantifind/wisp/pull/9#issuecomment-71212659>.
>



-- 
Austin Gibbons
Research
quantifind.com  |  @austinbg <https://twitter.com/austinbg>  |  *708 601
4894*


<http://quantifind.com/>

**Comment by AustinBGibbons** [comment]
I have made several changes:

- support key:value pairs and triplets through magnet pattern
- more accurately bin data (thanks @hudsonzhao)
- default numBins to the square-root of the number of datapoints
- Only use square-root number of labels on x-axis

File: core/src/main/scala/com/quantifind/charts/repl/BinnedData.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  def toBinned(): Iterable[(String, Double)] = coupledTripletBinned(data.map{case(a, b, c) => ((a, b), c)})
+}
+
+class CoupledTripletBinned[A, B, C: Numeric](data: Iterable[((A, B), C)]) extends BinnedData {
+  def toBinned(): Iterable[(String, Double)] = coupledTripletBinned(data)
+}
+
+class IterableBinned[A: Numeric](data: Iterable[A], numBins: Int = -1) extends BinnedData {
+
+  // Conditionally format on nuumber of leading zeros
```
Comment (ryanlecompte) [inline core/src/main/scala/com/quantifind/charts/repl/BinnedData.scala]: typo `nuumber`

File: core/src/main/scala/com/quantifind/charts/repl/BinnedData.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ * Date: 1/29/15
+ *
+ * Uses the magnet pattern to bin data, in preparation for building a histogram
+ */
+trait BinnedData {
+  def toBinned(): Iterable[(String, Double)]
+
+  def coupledTripletBinned[A, B, C: Numeric](data: Iterable[((A, B), C)]) = {
+    def numericToDouble[X](x: X)(implicit ev: Numeric[X]): Double = ev.toDouble(x)
+    data.map{case((a, b), c) => s"$a - $b" -> numericToDouble(c)}
```
Comment (squito) [inline core/src/main/scala/com/quantifind/charts/repl/BinnedData.scala]: you don't need the helper function to get a handle on the implicit `Numeric[C]`, you can do `implicitly[Numeric[C]]`, so you could do:

```
def coupledTripletBinned[A, B, C: Numeric](data: Iterable[((A, B), C)]) = {
  val numeric = implicitly[Numeric[C]]
  data.map{case((a, b), c) => s"$a - $b" -> numeric.toDouble(c)}
}
```

or

```
def coupledTripletBinned[A, B, C: Numeric](data: Iterable[((A, B), C)]) = {
  data.map{case((a, b), c) => s"$a - $b" -> implicitly[Numeric[C]].toDouble(c)}
}
```


or for that matter, the `C : Numeric` is really just syntactic sugar for this version, its got an identical signature:

```
def coupledTripletBinned[A, B, C](data: Iterable[((A, B), C)])(implicit ev: Numeric[C]) = {
  data.map{case((a, b), c) => s"$a - $b" -> ev.toDouble(c)}
}
```

**Comment by squito** [comment]
looks great!  would love to some test cases on the various binning, even just as a minimal usage example.  btw, I've heard of a little interest in this over here, before I even did any promotion of it

**Comment by AustinBGibbons** [comment]
Thanks for the feedback! I've changed the numeric evidence, add some test binning, and fixed the typo :-)

I'm going to merge this after it builds

**Pull request merged**